### PR TITLE
Move CLI completion generator to component-base.

### PR DESCRIPTION
**Pull request by jpeach**
**What type of PR is this?**

/kind cleanup

**What this PR does / why we need it**:

Both kubectl and kubeadm do roughly the same thing to generate
command-line completions, so move the common code to the component-base
repository. Rather than generating custom zsh completions, use the
default one from spf13/cobra supports zsh since v1.1.0.

**Special notes for your reviewer**:

We retain (at least for now) the command-specific help and error
handling since the command use different facilities for those.

**Does this PR introduce a user-facing change?**:
```release-note
NONE
```

### CODE REVIEW

**Comment by marckhouzam** [comment]
Be aware that Cobra's support for native zsh completion will not support the current custom completions that `kubectl` has implemented in bash script (https://github.com/spf13/cobra/blob/master/shell_completions.md#limitations).

`kubectl` must first transition all custom completion code from bash to Go, which is available in https://github.com/kubernetes/kubernetes/pull/96087

**Comment by jpeach** [comment]
> Be aware that Cobra's support for native zsh completion will not support the current custom completions that `kubectl` has implemented in bash script (https://github.com/spf13/cobra/blob/master/shell_completions.md#limitations).
> 
> `kubectl` must first transition all custom completion code from bash to Go, which is available in #96087

Thanks @marckhouzam, there's a lot of complexity that I missed here :)

Perhaps the right approach is (after #96087 lands) to make the same transition for Kubeadm. Once that's done, then we can look again at whether it is worth consolidating this stuff.

**Comment by fabriziopandini** [comment]
Interesting. 
I'm wondering If we should keep this in the radar for Cluster API/clustrctl as well which is implementing bash based shell completion in https://github.com/kubernetes-sigs/cluster-api/pull/3808
/cc @superbrothers

**Comment by neolit123** [comment]
i like the idea of consolidating.

> Perhaps the right approach is (after #96087 lands) to make the same transition for Kubeadm. Once that's done, then we can look again at whether it is worth consolidating this stuff.

SGTM.

**Review by neolit123 — COMMENTED** [review]
/ok-to-test
/priority backlog
/triage accepted

File: staging/src/k8s.io/component-base/cli/completion/completion.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+	"github.com/spf13/cobra"
+)
+
+var (
+	completionShells = map[string]func(out io.Writer, cmd *cobra.Command) error{
+		"bash": func(out io.Writer, cmd *cobra.Command) error {
+			return cmd.GenBashCompletion(out)
+		},
+		"zsh": func(out io.Writer, cmd *cobra.Command) error {
+			return cmd.GenBashCompletion(out)
```
Comment (superbrothers) [inline staging/src/k8s.io/component-base/cli/completion/completion.go:32]: Is this `GenZshCompletion`?

```suggestion
			return cmd.GenZshCompletion(out)
```

**Pull request closed without merging**